### PR TITLE
Introduce Vnet and nic

### PR DESCRIPTION
**Pull request by furkansahin**
This commit adds basic functionality to the controlplane to support
virtual subnets and virtual nics.

We are introducing two new Nexus files. They are responsible from the
lifecycle of virtual subnets and virtual nics.

What is a virtual subnet?
Virtual subnets are private subnets that are allocated to customers and
their individual services. They can create a private subnet and add
their resources into it to have a private networking setup. For now, we
allocate a subnet size of /26 for ipv4. This will give 64 distinct
private ipv4 addresses to the customer resources. We want to keep it
limited at the moment because we can create a limited number of
ipsec tunnels per vm. We make use of Class A, B, C private ipv4 address
spaces.

What is a virtual nic?
NIC (Network Interface Card) means that there is a virtual interface
that can be added to a VM that will give the ability to exist inside a
private network. The private IP addresses allocated in the virtual
subnet are assigned to VNics when they are created. Therefore, each Nic
has 1 private ipv4 and 1 private ipv6 address. When we assign a Nic to a
VM, this VM will be discoverable in the network that the nic is created
from.

After this commit, each VM provisioning will require either to create
a new subnet and a nic under it to be assigned to the VM, or we will get
the reference of an existing Nic to make use of it. For example;
```
    # let's create a subnet
    subnet_s = Prog::Vnet::SubnetNexus.assemble(nil, name: "vmfurkan-subnet",
                  location: "hetzner-hel1")

    # let's create a nic in it
    nic_s = Prog::Vnet::NicNexus.assemble(nil, subnet_s.id, name: "furkan-nic")

    # now, we can create a VM by using the nic id
    st = Prog::Vm::Nexus.assemble(mykey, nil, nic_id: nic_s.id)

    # or, we can create a vm by passing the subnet id and let it create
    # a new nic
    st = Prog::Vm::Nexus.assemble(mykey, nil, private_subnet_id: subnet_s.id)

    # or, we can simply choose to not pass anything, and it will create
    # both resources from scratch
    st = Prog::Vm::Nexus.assemble(mykey, nil)
```

So, by making use of an existing subnet_id to create new VMs, we can add
the into the same private network.

We also moved the IpsecTunnel <=> VMs relationship
to IpsecTunnel <=> Nics. Because the tunnels are simply used for private
networking. The addresses to create tunnels for are all stored in Nics,
so it makes more sense to move the relatipship as well. Now, with these
2 commits combined, refresh_mesh is handled differently;

1. When a new VM joins to a subnet or it leaves a subnet, we
incr_refresh_mesh for that subnet.
2. SubnetNexus picks up the notification and let's all the Nics on that
subnet know that there has to be a refresh operation.
3. NicNexus picks up the notification and calls ipsectunnel abjects
connected to it to perform a refresh.

- When leaving a network, the leaving nic entity has the responsibility to
let the subnet know about the situation.
- When joining a network, the resource (for now only vm) has the
responsibility to let the subnet know about the situation. 
The difference is because the nic may have been created much before but
not being used yet. It's only needed to refresh ipsec_tunnels when there is
a new resource making use of this nic.

There are some changes on the dataplane side as well. Most of the
work is to simply support the new parameter `nics` for prepvm. We
generate the tap device names, build the vm network configuration and
add the dhcp-range values for each private ip to the dnsmasq.conf.
We also switch to static mac addresses.

### CODE REVIEW

File: model/private_subnet.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+class PrivateSubnet < Sequel::Model
+  many_to_one :vm
+  one_to_many :nics, key: :private_subnet_id
+
+  PRIVATE_SUBNET_RANGES = [
+    "10.0.0.0/8",
+    "172.16.0.0/12",
+    "192.168.0.0/16"
+  ]
```
Comment (fdr) [inline model/private_subnet.rb:13]:
```suggestion
  ].freeze
```

File: migrate/20230711_add_nic.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      column :created_at, :timestamptz, null: false, default: Sequel.lit("now()")
+      column :private_ipv4, :inet, null: false
+      column :private_ipv6, :inet, null: false
+      column :vm_id, :uuid, null: false
+      column :encryption_key, :text, null: false
+    end
+
+    alter_table(:vm) do
+      add_foreign_key :nic_id, :nic, type: :uuid
+      add_foreign_key :private_subnet_id, :private_subnet, type: :uuid
```
Comment (fdr) [inline migrate/20230711_add_nic.rb:26]: isn't this duplicate/conflicts with multiple NICs in multiple subnets? seems like you might want neither column to do something like `Nic.where(vm_id: Vm.id)` (but Sequel's `many_to_one` association code should handle it for you)

**Comment by furkansahin** [comment]
I will still add tests.

**Comment by furkansahin** [comment]
This PR also has a huge test coverage improvement
Screenshot 2023-07-20 at 15 57 11

**Comment by fdr** [comment]
Unfortunately for you, you'll need to ubidify 1349ea6ac9f2dbcb27df9e966fa7613ffc526e8a

**Comment by fdr** [comment]
I pre-endorse this patch in its major theory, it would be nice if boring hunks were committed separately (like moving code blocks with minor or no alteration) so the interesting stuff were uninterrupted.

Depending on how I feel about overall size or longevity of the larger patch, I sometimes send those out mentioning that it's part of completing an objective and has minor or no expected behavior change to get it merged slightly faster. Usually anyone is capable of quickly skimming and approving those to get it integrated and avoid conflicts and ease rebasing.

I will be out of pocket on Friday so I'd advise you to find another reviewer as to avoid delay.

I may come back with more comments tonight, I have an engagement so it's a bit unclear.

Thanks for your quick work on getting this data model in; I think it will spare us at least a week in the next quarter (imagining renaming or backwards-compatible code for TAP, and introducing nics to a live system for the first time)

File: prog/vnet/nic_nexus.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    # maybe wait vnet
+    subnet = PrivateSubnet[private_subnet_id]
+    unless subnet
+      fail "Given subnet doesn't exist with the id #{private_subnet_id}"
+    end
+
+    ubid = UBID.generate(UBID::TYPE_NIC)
+    tap_name = Nic.ubid_to_tap_name(ubid)
+
+    Validation.validate_name(tap_name)
```
Comment (pykello) [inline prog/vnet/nic_nexus.rb:21]: Not sure if we need this, since this is not a value entered by user (unlike the `name` in `Nexus.assemble` which could be user entered)

Did you mean to validate the "name" parameter instead?

Also what happens if `name` is `nil`? Do we keep is as `nil` or `||` with the auto-generated `tap_name` similar to `Nexus.assemble`?

File: prog/vnet/subnet_nexus.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    nap 1
+  end
+
+  def destroy
+    if ps.nics.any? { |n| !n.vm_id.nil? }
+      fail "Cannot destroy subnet with active nics, first clean up the attached resources"
+    end
+
+    if ps.nics.empty?
+      ps.delete
```
Comment (enescakir) [inline prog/vnet/subnet_nexus.rb:63]: Should we do any cleanup at dataplane?

File: prog/vnet/nic_nexus.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    end
+
+    decr_refresh_mesh
+    hop :wait
+  end
+
+  def destroy
+    DB.transaction do
+      nic.src_ipsec_tunnels_dataset.delete
+      nic.dst_ipsec_tunnels_dataset.delete
```
Comment (enescakir) [inline prog/vnet/nic_nexus.rb:79]: Should we do any cleanup at dataplane? How is deleted ipsec tunnels cleaned up?

File: prog/vm/nexus.rb [hunk 6]
```diff
@@ -359,18 +310,19 @@ def destroy
     end
 
     DB.transaction do
-      vm.private_subnets_dataset.delete
+      vm.nics.map do |nic|
+        nic.update(vm_id: nil)
```
Comment (enescakir) [inline prog/vm/nexus.rb:305]: When an user created an VM and delete it without knowing anything about networking, he will have leftover nic and subnet.

File: prog/vnet/subnet_nexus.rb (hunk 4 above)
Comment (furkansahin) [reply prog/vnet/subnet_nexus.rb:63]: Nope, private subnets are just logically existing entities.

File: prog/vnet/nic_nexus.rb (hunk 5 above)
Comment (furkansahin) [reply prog/vnet/nic_nexus.rb:79]: Since Nics are detached/destroyed only when VM gets destroyed, we don't have to clean-up anything.

File: prog/vm/nexus.rb [hunk 7]
```diff
@@ -29,18 +29,44 @@ def self.assemble(public_key, project_id, name: nil, size: "m5a.2x",
...
+      # private_subnet_id.
+      # - If the user did not provide nic_id but the private_subnet_id, that private_subnet
+      # must exist, otherwise we fail.
+      # - If the user did not provide nic_id but the private_subnet_id and that subnet exists
+      # then we create a nic on that subnet.
+      # - If the user provided neither nic_id nor private_subnet_id, that's OK, we create both.
+      nic = nil
+      subnet = nil
+      if nic_id
+        nic = Nic[nic_id]
```
Comment (byucesoy) [inline prog/vm/nexus.rb:45]: What happens if user passes id of another customer?

File: spec/coverage_helper.rb [hunk 8]
```diff
@@ -5,7 +5,7 @@
 
   SimpleCov.start do
     enable_coverage :branch
-    minimum_coverage line: 90.45, branch: 91
+    minimum_coverage line: 94, branch: 94.2
```
Comment (byucesoy) [inline spec/coverage_helper.rb:8]: Commit message needs an update :(

File: spec/ubid_spec.rb [hunk 9]
```diff
@@ -211,11 +220,12 @@ def string_kv(obj)
     sv = VmStorageVolume.create_with_id(vm_id: vm.id, size_gib: 5, disk_index: 0, boot: false)
     kek = StorageKeyEncryptionKey.create_with_id(algorithm: "x", key: "x", init_vector: "x", auth_data: "x")
     account = Account.create_with_id(email: "x@y.net")
-    project = account.create_project_with_default_policy("x")
-    policy = project.access_policies.first
-    atag = AccessTag.create_with_id(project_id: project.id, hyper_tag_table: "x", name: "x")
-    tun = IpsecTunnel.create_with_id(src_vm_id: vm.id, dst_vm_id: vm.id)
-    subnet = VmPrivateSubnet.create_with_id(vm_id: vm.id, net6: "0::0", net4: "127.0.0.1")
+    prj = account.create_project_with_default_policy("x")
```
Comment (byucesoy) [inline spec/ubid_spec.rb:223]: Are these left over from rebase?

File: prog/vm/nexus.rb (hunk 7 above)
Comment (furkansahin) [reply prog/vm/nexus.rb:45]: That's a good question, though, right now, we don't have much mechanism around entity ownership tied to a customer (organisation?). We have projects and subnets are tied to a project, so I think I can use it. The https://github.com/ubicloud/ubicloud/pull/334 actually implements something similar, I will get that pr in shortly after this one

Comment (fdr) [reply prog/vm/nexus.rb:45]: well, we don't have entity ownership as such because we don't have an owner role per se, because we're ABAC. This may be an @enescakir kind of question, though it should prompt survey of comparison implementations (the hyperscalers).

We may not do such authorizations in nexuses anyway, it may live in the API, though I sort of think defense-in-depth/avoiding horrible operator mistakes even at the cost of re-checking authorization is a good idea.

File: prog/vm/nexus.rb [hunk 10]
```diff
@@ -177,6 +191,10 @@ def allocate
   def start
     register_deadline(:wait, 10 * 60)
 
+    unless vm.nics.all? { |n| n.strand.label == "wait" }
+      nap 10
```
Comment (fdr) [inline prog/vm/nexus.rb:187]: Put a pin in this for later: we should probably do a `donate` variant here, even though nic strands are not children, rather than waiting for the stars (or nap loops) to align.

Comment (fdr) [reply prog/vm/nexus.rb:187]: Given @byucesoy is kinda working his way into the strand system, I'm going to weigh it into his consciousness, too.

File: prog/vnet/nic_nexus.rb [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+  def self.assemble(private_subnet_id, name: nil, ipv6_addr: nil, ipv4_addr: nil)
+    unless (subnet = PrivateSubnet[private_subnet_id])
+      fail "Given subnet doesn't exist with the id #{private_subnet_id}"
+    end
+
+    ubid = UBID.generate(UBID::TYPE_NIC)
+    name ||= Nic.ubid_to_name(ubid)
+
+    Validation.validate_name(name)
```
Comment (fdr) [inline prog/vnet/nic_nexus.rb:14]: Interesting, should this be pushed down in a separate patch? are there situations we don't want a "valid" name?

File: prog/vnet/nic_nexus.rb [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    "0x" + SecureRandom.bytes(36).unpack1("H*")
+  end
+
+  def nic
+    @nic ||= Nic[strand.id]
+  end
+
+  def wait_vnet
+    unless nic.private_subnet.strand.label == "wait"
+      nap 1
```
Comment (fdr) [inline prog/vnet/nic_nexus.rb:38]: 'nother one of these loops (not your fault, but it suggests we lack a programming model we need)

File: model/ipsec_tunnel.rb [hunk 13]
```diff
@@ -7,4 +7,58 @@ class IpsecTunnel < Sequel::Model
...
+      "#{dst_private_addr_4} #{src_direction} " \
+      "#{spi} #{spi4} #{key}")
+
+    # setup destination ipsec tunnels
+    dst_nic.vm.vm_host.sshable.cmd("sudo bin/setup-ipsec " \
+      "#{dst_namespace} #{src_clover_ephemeral} " \
+      "#{dst_clover_ephemeral} #{src_private_addr_6} " \
+      "#{dst_private_addr_6} #{src_private_addr_4} " \
+      "#{dst_private_addr_4} #{dst_direction} " \
+      "#{spi} #{spi4} #{key}")
```
Comment (fdr) [inline model/ipsec_tunnel.rb:46]: I find it interesting that bilateral set-up (where we must shell into one machine and then the other in one shot) survives another day. I don't have deep thoughts about it yet, except that it requires deep thought.

File: rhizome/lib/vm_setup.rb [hunk 14]
```diff
@@ -173,20 +186,28 @@ def routes4(ip4, ip4_local)
...
   end
 
-  def nat4(ip4, private_ipv4)
+  def nat4(ip4, nics)
     return unless ip4
     public_sub = NetAddr::IPv4Net.parse(ip4)
     public_ipv4 = public_sub.network.to_s
 
+    # TODO: this needs to be fixed, the public to private NAT will be an issue??
+    # maybe it's just OK?
```
Comment (fdr) [inline rhizome/lib/vm_setup.rb:209]: can we resolve this question?

File: rhizome/lib/vm_setup.rb [hunk 15]
```diff
@@ -487,9 +524,13 @@ def forwarding
     r("ip netns exec #{q_vm} sysctl -w net.ipv4.ip_forward=1")
   end
 
-  def install_systemd_unit(max_vcpus, cpu_topology, mem_gib, vhost_sockets)
+  def install_systemd_unit(max_vcpus, cpu_topology, mem_gib, vhost_sockets, nics)
     cpu_setting = "boot=#{max_vcpus},topology=#{cpu_topology}"
 
+    tapnames = nics.map { |net6, net4, tap, mac|
+      "-i #{tap} "
```
Comment (fdr) [inline rhizome/lib/vm_setup.rb:531]: Is this a subtle way to get a trailing space of a bit of extra safety when being spliced into a command line? Otherwise you could use `join(' ')`

File: rhizome/lib/vm_setup.rb [hunk 16]
```diff
@@ -542,7 +587,7 @@ def install_systemd_unit(max_vcpus, cpu_topology, mem_gib, vhost_sockets)
 --console off --serial file=#{vp.serial_log} \
 --cpus #{cpu_setting} \
 --memory size=#{mem_gib}G,hugepages=on,hugepage_size=1G \
---net "mac=#{guest_mac},tap=tap#{@vm_name},ip=,mask="
+#{net_params.join("\n")}
```
Comment (fdr) [inline rhizome/lib/vm_setup.rb:590]: seems like you want a line continuation mark (`\`) in the join, not just newline.

**Review by fdr — COMMENTED** [review]
5479bac741aae3e1fff098cbc01a4cde274ebe6cs commit message clock block has indentation issues.

it's complicated work, and a bit step forward. I mostly had comments. I think the most serious maybe-bug I found was the line continuation issue.

File: rhizome/lib/vm_setup.rb (hunk 14 above)
Comment (furkansahin) [reply rhizome/lib/vm_setup.rb:209]: It's just OK for now, since we have 1 nic attached atm. So the NAT can be performed from public ip to whatever the first NIC has as a private ip.

File: prog/vnet/nic_nexus.rb (hunk 11 above)
Comment (furkansahin) [reply prog/vnet/nic_nexus.rb:14]: Yeah, characters in resource name may result in our paths in the console not usable. I can leave this to the UI PR though.

File: prog/vm/nexus.rb (hunk 7 above)
Comment (enescakir) [reply prog/vm/nexus.rb:45]: Private subnets are associated with projects. So we can find nic in nics of private subnets of project instead of all nics.

Previous resources (vm, project) do authorization checks at routes/controller level. It blocks customers to access other customers resources, but not operator.

If we want to avoid operator mistakes and move some kind of authorization check in progs, we need to think more about it. Because we need a subject for authorization checks. It's customer at routes/controllers. What will be subject in progs? Because operator doesn't have permissions to access objects. So our current authorization code doesn't work without a subject.

Comment (furkansahin) [reply prog/vm/nexus.rb:45]: @enescakir can you check the latest version of the operation here.

File: rhizome/lib/vm_setup.rb (hunk 14 above)
Comment (furkansahin) [reply rhizome/lib/vm_setup.rb]: I think it's even fine when we start supporting multiple nics.

File: rhizome/lib/vm_setup.rb (hunk 15 above)
Comment (furkansahin) [reply rhizome/lib/vm_setup.rb]: changed it as suggested

File: rhizome/lib/vm_setup.rb (hunk 16 above)
Comment (furkansahin) [reply rhizome/lib/vm_setup.rb]: Thanks for catching this. I corrected it.

File: prog/vm/nexus.rb (hunk 7 above)
Comment (byucesoy) [reply prog/vm/nexus.rb:45]: It seems avoiding operator mistakes requires significant investment right now. I think similar type of investment might be needed at the time of building an admin panel. I'd suggest leaving that to there. If we can check the authorization in the API level to prevent potential vulnerability, it is a win.

File: prog/vm/nexus.rb (hunk 10 above)
Comment (byucesoy) [reply prog/vm/nexus.rb:191]: You mean something like donate to non-child strand? Sure that would be useful. I'll think about it.

File: prog/vm/nexus.rb (hunk 7 above)
Comment (enescakir) [reply prog/vm/nexus.rb:45]: Yeah. I think it's better now. We need to add extra authorization check in you UI PR.

Comment (furkansahin) [reply prog/vm/nexus.rb:45]: Thanks
